### PR TITLE
[KYUUBI #5010] Make Kyuubi server's connection URL configurable

### PR DESCRIPTION
**Pull request by touchida**
### _Why are the changes needed?_

This PR resolves https://github.com/apache/kyuubi/issues/5010, by introducing a new configuration property that makes it possible to override the hostname or ip of the Kyuubi server's thrift frontend service that will be registered for the service discovery ensemble.

### _How was this patch tested?_
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [X] Add screenshots for manual tests if appropriate

Before applying the patch, when we set `kyuubi.frontend.bind.host=0.0.0.0`, the znode becomes as follows.
```
[zk: some_zk_host:some_zk_port(CONNECTED) 0] ls /some_kyuubi_zk_namespace
[serviceUri=0.0.0.0:10009;version=1.8.0-SNAPSHOT;sequence=0000000007]
```
After applying the patch, when we set `kyuubi.frontend.bind.host=0.0.0.0` and `kyuubi.frontend.advertised.host=some_kyuubi_host`, it becomes as follows.
```
[zk: some_zk_host:some_zk_port(CONNECTED) 0] ls /some_kyuubi_zk_namespace
[serviceUri=some_kyuubi_host:10009;version=1.8.0-SNAPSHOT;sequence=0000000009]
```

- [X] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

`./build/mvn clean install -pl kyuubi-common` succeeded.

### CODE REVIEW

**Review by cxzl25 — APPROVED** [review]

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -507,6 +507,15 @@ object KyuubiConf {
       .serverOnly
       .fallbackConf(FRONTEND_BIND_PORT)
 
+  val FRONTEND_THRIFT_CONNECTION_URL_HOST: ConfigEntry[String] =
+    buildConf("kyuubi.frontend.thrift.connection.url.host")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:511]: The following configuration is used by Apache Kafka

```
advertised.host.name
advertised.port
advertised.listeners
```

I would call it `kyuubi.frontend.thrift.advertised.host`, if you want, we call it `kyuubi.frontend.advertised.host` and apply it to all frontends.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 2]
```diff
@@ -507,6 +507,15 @@ object KyuubiConf {
...
+      .doc("Hostname or IP of the Kyuubi server's thrift frontend service that will be " +
+        "registered for the service discovery ensemble.")
+      .version("1.8.0")
+      .serverOnly
+      .stringConf
+      .createWithDefault("")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:517]: how about using `OptionalConfigEntry` and `createOptional` instead?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 1 above)
Comment (touchida) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: Thanks for your advice!
`advertised.host` sounds better to me.
Fixed at 34c61d9.

I also applied the configuration property to all frontends.
But I did not introduce separate properties for each frontend at this time.
It's possible, but I think such a use case is rare.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 2 above)
Comment (touchida) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: Fixed at 34c61d9.

**Comment by touchida** [comment]
The following failed test will not be related to my PR, and it succeeds in my local environment.
```
DockerizedZkServiceDiscoverySuite:
- publish instance to discovery service
- KYUUBI #304: Stop engine service gracefully when related node is deleted
- distribute lock *** FAILED ***
  lockLatch.await(5000L, MILLISECONDS) was false (DiscoveryClientTests.scala:146)
- getAndIncrement method test
- setData method test
- acl for zookeeper
- set up zookeeper auth
- stop engine in time while zk ensemble terminates
- watcher for zookeeper
Run completed in 47 seconds, 328 milliseconds.
Total number of tests run: 9
Suites: completed 2, aborted 0
Tests: succeeded 8, failed 1, canceled 0, ignored 0, pending 0
*** 1 TEST FAILED ***
```

@pan3793
Could you kindly review my PR again when you have time?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTrinoFrontendService.scala [hunk 3]
```diff
@@ -46,19 +46,26 @@ class KyuubiTrinoFrontendService(override val serverable: Serverable)
...
 
   override def connectionUrl: String = {
     checkInitialized()
-    server.getServerUri
+    val advertisedHost = conf.get(FRONTEND_ADVERTISED_HOST)
+    if (advertisedHost.isEmpty) {
+      server.getServerUri
+    } else {
+      s"${advertisedHost.get}:$port"
+    }
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTrinoFrontendService.scala:68]:
```suggestion
    conf.get(FRONTEND_ADVERTISED_HOST) match {
      case Some(advertisedHost) => s"$advertisedHost:$port"
      case None => server.getServerUri
    }
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiRestFrontendService.scala [hunk 4]
```diff
@@ -68,19 +68,26 @@ class KyuubiRestFrontendService(override val serverable: Serverable)
...
-      conf.get(FRONTEND_REST_BIND_PORT),
+      port,
       conf.get(FRONTEND_REST_MAX_WORKER_THREADS))
     super.initialize(conf)
   }
 
   override def connectionUrl: String = {
     checkInitialized()
-    server.getServerUri
+    val advertisedHost = conf.get(FRONTEND_ADVERTISED_HOST)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiRestFrontendService.scala:85]: ditto, replace with pattern match

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala [hunk 5]
```diff
@@ -94,7 +94,12 @@ class KyuubiMySQLFrontendService(override val serverable: Serverable)
 
   override def connectionUrl: String = {
     checkInitialized()
-    s"${serverAddr.getCanonicalHostName}:$port"
+    val advertisedHost = conf.get(FRONTEND_ADVERTISED_HOST)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala:97]: ditto

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/TFrontendService.scala [hunk 6]
```diff
@@ -112,12 +112,17 @@ abstract class TFrontendService(name: String)
...
+      serverHost match {
+        case Some(h) => h // respect user's setting ahead
+        case None if conf.get(FRONTEND_CONNECTION_URL_USE_HOSTNAME) =>
+          serverAddr.getCanonicalHostName
+        case None =>
+          serverAddr.getHostAddress
+      }
+    } else {
+      advertisedHost.get
     }
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/TFrontendService.scala:126]:
```suggestion
    (conf.get(FRONTEND_ADVERTISED_HOST), serverHost) match {
      case (Some(advertisedHost), _) => advertisedHost
      case (None, Some(h)) => h
      case (None, None) if conf.get(FRONTEND_CONNECTION_URL_USE_HOSTNAME) =>
        serverAddr.getCanonicalHostName
      case (None, None) => serverAddr.getHostAddress
    }
```

**Review by pan3793 — APPROVED** [review]
thank you @touchida. LGTM, except for code style,

**Comment by touchida** [comment]
@pan3793 
Thanks for your careful review!
Let me fix the points you suggested soon.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 7]
```diff
@@ -429,6 +429,17 @@ object KyuubiConf {
     .stringConf
     .createOptional
 
+  val FRONTEND_ADVERTISED_HOST: OptionalConfigEntry[String] =
+    buildConf("kyuubi.frontend.advertised.host")
+      .doc(
+        "Hostname or IP of the Kyuubi server's frontend services to publish to " +
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:435]:
```suggestion
      .doc("Hostname or IP of the Kyuubi server's frontend services to publish to " +
```

**Review by pan3793 — APPROVED** [review]

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 8]
```diff
@@ -429,6 +429,16 @@ object KyuubiConf {
     .stringConf
     .createOptional
 
+  val FRONTEND_ADVERTISED_HOST: OptionalConfigEntry[String] =
+    buildConf("kyuubi.frontend.advertised.host")
+      .doc("Hostname or IP of the Kyuubi server's frontend services to publish to " +
+          "external systems such as the service discovery ensemble and metadata store. " +
+          "Use it when you want to advertise a different hostname or IP than the bind host.")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:436]:
```suggestion
        "external systems such as the service discovery ensemble and metadata store. " +
        "Use it when you want to advertise a different hostname or IP than the bind host.")
```

**Comment by pan3793** [comment]
Thanks, merged to master

**Comment by touchida** [comment]
@pan3793 @cxzl25 
Thanks for your review!

**Pull request closed without merging**